PocketMine-MP.phar self-extraction to decompressed cache

dktapps opened the pull request:
Because ext-phar sucks, tmp gets spammed by cache files for every thread when loading files from the phar on the fly.

Instead, we convert the `.phar` into a decompressed `.tar` in the tmp directory and require files from inside it. Surprisingly, this works because `ext-phar` supports `tar` and `zip` natively. No stream wrapper is required, as the `PocketMine.php` bootstrap loads files relative to its location, so the cache is automatically used for everything.

To be honest I would rather get rid of phars entirely, but they are still the easiest way to have PhpStorm load PocketMine-MP API information for now, and the alternatives are more complicated and inconvenient.

### Caveats
Everywhere that previously used `new Phar(Phar::running(false))` in the core code needs to be updated to use `PharData` for this to work correctly. Plugins don't need to do anything.

### Why not just use `Phar::decompressFiles()`?
This requires setting `phar.readonly` to `0`, which is a security issue. Technically, we could have used a subprocess to do this, but it just didn't seem right.

### WTF? `phar://` can be used on `tar` files???
Yup. I was just as surprised to find out that `require` works in such contexts.

### Relevant issues
- Closes #6214 

## Changes
### API changes
None.

### Behavioural changes
Server startup will be slightly slower, as the phar has to decompress and convert itself into a `.tar`. However, testing showed that this generally takes less than 200 ms, so it should be barely noticeable.

## Backwards compatibility
No BC issues.

## Tests
Locally tested and the CI will also verify

dktapps committed:
PocketMine-MP.phar now self-extracts to a cache dir, instead of loading files from the phar on the fly
Because ext-phar sucks, tmp gets spammed by cache files for every thread when loading files from the phar on the fly.
Instead, we extract the phar to a hidden cache directory, avoiding the problem entirely.
It's not a great solution, since removing the cache can be slow (especially on Windows), but it's better than flooding tmp with garbage.

To be honest I would rather get rid of phars entirely, but they are still the easiest way to have PhpStorm load PocketMine-MP API information for now, and the alternatives are more complicated and inconvenient.

SenseiTarzan commented:
I have a client who had 16 GB in tmp which came from pmmp because when you go into it he had a whole file which starts with php and with pmmp inside it decompresses into a single file

SenseiTarzan commented:
I have a client who had 10 GB in tmp which came from pmmp because when you go into it he had a whole file which starts with php and with pmmp inside it decompresses into a single file XD

SOF3 commented on src/PocketMine.php:
```diff
@@ -175,9 +175,6 @@ function emit_performance_warnings(\Logger $logger){
 			$logger->warning("Debugging assertions are enabled. This may degrade performance. To disable them, set `zend.assertions = -1` in php.ini.");
 		}
-		if(\Phar::running(true) === ""){
-			$logger->warning("Non-packaged installation detected. This will degrade autoloading speed and make startup times longer.");
```
rip phar

dktapps replied:
To be honest I think this was always a lie. Decompressing the phar contents negates any gain had by packing everything into one file. The only way this warning would have been valid is if the phar had zero compression.

dries-c replied:
It was probably manufactured during the time where releases only had function imports.

dktapps replied:
I'd forgotten the phar releases used to be preprocessed. You might be right about that.

SOF3 replied:
I thought windows had better performance loading from a single archive.

dktapps replied:
Like I said, the I/O overhead is offset by decompression overhead. There's little observable difference in practice.

On *nix src is pretty much always going to be faster, and that's where the majority of the userbase is.

SOF3 commented on build/server-phar.php:
```diff
@@ -159,21 +161,7 @@ function main() : void{
-__HALT_COMPILER();
-STUB
-,
+		Filesystem::fileGetContents(Path::join(__DIR__, 'server-phar-stub.php')) . "\n__HALT_COMPILER();",
```
why not write this inside server-phar-stub.php?

dktapps replied:
No particular reason, I just felt it looked weird to have __HALT_COMPILER in a script file
I also don't know what effects trailing newlines might have

SOF3 commented on build/server-phar-stub.php:
```diff
@@ -0,0 +1,73 @@
+	}
+}
+
+$sourceDir = getcwd() . "/.PocketMine-MP.phar/";
```
Why don't we use tempdir here? Classes are still loaded from the same tempdir since we won't be including the phar again in threads. You can copy with mode 0o400 if you are concerned about other users tampering the file.

dktapps replied:
Multiple servers might want to extract different versions of the server at once. Having the cache locally avoids the need to add discriminators to separate them

SOF3 replied:
I am mainly concerned that users are under the impression they can directly modify the source code here.

dktapps replied:
Yeah, the same thing occurred to me, though I'm not sure how much we need to worry about this.

dktapps replied:
I'll also mention it's a pretty big pain to reliably differentiate these if we start going into /tmp. We'd need to derive the tmp discriminator from the PocketMine-MP.phar path to ensure reuse and proper cleanup on crash.

dktapps replied:
No longer applicable with latest revision

SOF3 commented on build/server-phar-stub.php:
```diff
@@ -0,0 +1,73 @@
+	if($prepare){
+		$phar = new \Phar(__FILE__);
+		echo "Preparing source cache..." . PHP_EOL;
+		$phar->extractTo($sourceDir);
```
actually we could chmod(400) all files inside the phar too so that users don't try to modify this file https://www.php.net/manual/en/pharfileinfo.chmod.php

dktapps replied:
No longer applicable on latest revision

dktapps commented:
My main sticking point with this change is extracting source files into a directory tree. This makes for a cache that's slow to clean up and validate. It would be better if we could have a single-file cache similar to ext-phar for this purpose (but obviously one that can be shared by multiple threads).

The only ways I can see to do this are:
- Create a second temporary phar with all files decompressed - problematic because of `phar.readonly`
- Add a stream wrapper for `phar://` to access a custom cache file, overriding `ext-phar`'s shitty cache mechanism - potentially problematic because it's difficult to ensure that all threads have the stream wrapper registered

I'm wondering if we could do something with Composer bootstrap files to auto-register phar stream wrappers to do this. If we can find a way to piggyback onto the autoloader, the threading problems would go away.

dktapps committed:
Merge remote-tracking branch 'origin/minor-next' into phar-self-extract

dktapps committed:
A cursed hack using temporary .tar files
this is far superior to directory extraction, being much faster to delete (and to verify if we decided to make the cache reusable).
I opted for a non-reusable cache in this instance as it didn't seem necessary. I only went for reusable cache previously due to the large startup/shutdown performance hit of using directories.

dktapps commented:
I managed to find a simpler way to do this with a single file, allowing us to avoid potentially fragile cache reuse across multiple runs.

Essentially, we create a .tar from the .phar using `Phar::convertToData()` with `compression: Phar::NONE` (which isn't restricted by `phar.readonly`), and then `require` files from inside the .tar using the `phar://` wrapper (yes, this actually works).

This approach requires changing any places that used `new Phar(Phar::running(false))` as they'll need to use `PharData` if the server is running from a cache.

dktapps committed:
tmpfiles on linux don't have an extension, so phar:// gets mad :(

dktapps committed:
Namespace for phar stub (to avoid polluting global namespace with function declarations)

dktapps commented on build/server-phar-stub.php:
```diff
@@ -0,0 +1,72 @@
+echo "Cache ready at $cacheName in " . number_format((hrtime(true) - $start) / 1e9, 2) . "s\n";
+register_shutdown_function(static function() use ($cacheName, $tmpReserveName) : void{
+	if(is_file($cacheName)){
+		\Phar::unlinkArchive($cacheName);
```
Unsure what effects this might have during crashes - I'm not sure if this handler will be called before or after `Server::crashDump`

dktapps replied:
Turns out it's called before, which is quite annoying :(

dktapps commented:
We also need a cache cleanup mechanism, for when servers crash and don't clean up after themselves

dktapps committed:
Merge branch 'minor-next' into phar-self-extract

dktapps committed:
Phar cache is now cleaned up by new server instances on startup
this ensures that dead caches are properly removed, regardless of what caused the termination of the previous process.

dktapps commented:
This now includes a cache cleanup mechanism, amongst other improvements (and still being littered with debug messages)

dktapps committed:
Beware race condition when scanning for tmp files to clean up
